geode-core/src/test/java/org/apache/geode/test/dunit/cache/rules/tests/CacheRuleTest.java

```diff
@@ -0,0 +1,287 @@
+  }
+
+  @Test
+  public void createCacheInAllCreatesCluster() throws Exception {
+    runTestWithValidation(CreateCacheInAllCreatesCluster.class);
+  }
+
+  @Test
+  public void addConfigAffectsAllVMs() throws Exception {
+    runTestWithValidation(AddConfigAffectsAllVMs.class);
+  }
+
+  @Test
+  public void replaceConfigCreatesLonersInAll() throws Exception {
+    runTestWithValidation(ReplaceConfigCreatesLonersInAll.class);
+  }
+
+  /**
+   * Used by test {@link #defaultDoesNothing()}.
+   */
+  public static class DefaultDoesNothing implements Serializable {
+
+    @Rule
+    public CacheRule cacheRule = CacheRule.builder().build();
+
+    @Test
+    public void getCache_returnsNullInAllVMs() throws Exception {
+      assertThat(cacheRule.getCache()).isNull();
+      for (VM vm : Host.getHost(0).getAllVMs()) {
+        vm.invoke(() -> assertThat(cacheRule.getCache()).isNull());
+      }
+    }
+
+    @Test
+    public void getCacheSingleton_returnsNullInAllVMs() throws Exception {
+      assertThat(GemFireCacheImpl.getInstance()).isNull();
+      for (VM vm : Host.getHost(0).getAllVMs()) {
+        vm.invoke(() -> assertThat(GemFireCacheImpl.getInstance()).isNull());
+      }
+    }
+  }
+
+  /**
+   * Used by test {@link #createCacheInLocal()}.
+   */
+  public static class CreateCacheInLocal implements Serializable {
+
+    @Rule
+    public CacheRule cacheRule = CacheRule.builder().createCacheInLocal().build();
+
+    @Test
+    public void getCache_returnsCacheInLocalOnly() throws Exception {
+      assertThat(cacheRule.getCache()).isNotNull();
+      for (VM vm : Host.getHost(0).getAllVMs()) {
+        vm.invoke(() -> assertThat(cacheRule.getCache()).isNull());
+      }
+    }
+  }
+
+  /**
+   * Used by test {@link #createCacheInOneVM()}.
+   */
+  public static class CreateCacheInOneVM implements Serializable {
+
+    @Rule
+    public CacheRule cacheRule =
+        CacheRule.builder().createCacheIn(Host.getHost(0).getVM(0)).build();
+
+    @Test
+    public void getCache_returnsCacheInOneVM() throws Exception {
+      assertThat(cacheRule.getCache()).isNull();
+
+      Host.getHost(0).getVM(0).invoke(() -> {
+        assertThat(cacheRule.getCache()).isNotNull();
+      });
+
+      for (int i = 1; i < Host.getHost(0).getVMCount(); i++) {
+        Host.getHost(0).getVM(i).invoke(() -> {
+          assertThat(cacheRule.getCache()).isNull();
+        });
+      }
+    }
+  }
+
+  /**
+   * Used by test {@link #createCacheInTwoVMs()}.
+   */
+  public static class CreateCacheInTwoVMs implements Serializable {
+
+    @Rule
+    public CacheRule cacheRule = CacheRule.builder().createCacheIn(Host.getHost(0).getVM(1))
+        .createCacheIn(Host.getHost(0).getVM(3)).build();
+
+    @Test
+    public void getCache_returnsCacheInTwoVMs() throws Exception {
+      assertThat(cacheRule.getCache()).isNull();
+
+      for (int i = 0; i < Host.getHost(0).getVMCount(); i++) {
+        if (i == 1 || i == 3) {
```

Comment:
Why not use VM2 instead (would simplify this logic)?

Reply:
You mean VM0 and VM2? The point was to have half of the VMs with Caches and half without. I also wanted them to be non-sequential VMs. The only other way to achieve that would be:
```java
if (i == 0 II i == 2)
```
Or I could use mod or other math conditional to get true if odd or even.

Reply:
Yeah, if you want the VMs to be non-sequential, this looks good.